src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/validation/SimpleSwitchRuleConverter.java

```diff
@@ -271,180 +233,179 @@ public List<SimpleSwitchRule> buildYFlowTransitSimpleSwitchRules(List<SimpleSwit
 
-        Map<Integer, List<GroupBucket>> groupMap = new HashMap<>();
-        if (groups != null && groups.getGroupEntries() != null) {
-            for (GroupEntry groupEntry : groups.getGroupEntries()) {
-                groupMap.put(groupEntry.getGroupId(), groupEntry.getBuckets());
-            }
-        }
 
         List<SimpleSwitchRule> simpleRules = new ArrayList<>();
-        for (FlowEntry flowEntry : rules.getFlowEntries()) {
-            simpleRules.add(buildSimpleSwitchRule(rules.getSwitchId(), flowEntry, meterMap, groupMap));
+        for (FlowSpeakerData flowSpeakerData : rules) {
+            simpleRules.add(buildSimpleSwitchRule(flowSpeakerData, meterMap, groupMap));
         }
         return simpleRules;
     }
 
+    private <T extends SpeakerData> Map<SwitchId, List<T>> filterSpeakerDataAndGroupBySwitchId(
+            List<SpeakerData> speakerData, Class<T> clazz) {
+
+        return speakerData.stream()
+                .filter(Objects::nonNull)
+                .filter(clazz::isInstance)
+                .map(clazz::cast)
+                .collect(Collectors.groupingBy(SpeakerData::getSwitchId));
+    }
+
     private SimpleSwitchRule buildSimpleSwitchRule(
-            SwitchId switchId, FlowEntry flowEntry, Map<Long, MeterEntry> meterMap,
-            Map<Integer, List<GroupBucket>> groupMap) {
+            FlowSpeakerData flowEntry, Map<Long, List<MeterSpeakerData>> meterMap,
+            Map<Integer, List<GroupSpeakerData>> groupMap) {
+
         SimpleSwitchRule rule = SimpleSwitchRule.builder()
-                .switchId(switchId)
-                .cookie(flowEntry.getCookie())
+                .switchId(flowEntry.getSwitchId())
+                .cookie(flowEntry.getCookie().getValue())
                 .pktCount(flowEntry.getPacketCount())
                 .byteCount(flowEntry.getByteCount())
-                .version(flowEntry.getVersion())
+                .version(flowEntry.getOfVersion().toString())
                 .build();
 
         if (flowEntry.getMatch() != null) {
-            rule.setInPort(NumberUtils.toInt(flowEntry.getMatch().getInPort()));
-            rule.setInVlan(NumberUtils.toInt(flowEntry.getMatch().getVlanVid()));
-            rule.setTunnelId(Optional.ofNullable(flowEntry.getMatch().getTunnelId())
-                    .map(Integer::decode)
-                    .orElse(NumberUtils.INTEGER_ZERO));
+            rule.setInPort(flowEntry.getMatch().stream().filter(e -> e.getField().equals(Field.IN_PORT))
+                    .map(FieldMatch::getValue).findFirst().orElse(NumberUtils.LONG_ZERO).intValue());
+
+            rule.setInVlan(flowEntry.getMatch().stream().filter(e -> e.getField().equals(Field.VLAN_VID))
+                    .map(FieldMatch::getValue).findFirst().orElse(NumberUtils.LONG_ZERO).intValue());
+
+            rule.setTunnelId(flowEntry.getMatch().stream().filter(e -> e.getField().equals(Field.OVS_VXLAN_VNI)
```

Comment:
why `e`?

Reply:
sorry, I will fix it. Totally agree that we have to use self explaining vars.. 

if you are curious why, because for example when we use for loop, we usually use "i" for iterator.

In Java's forEach() loop, the variable e is commonly used as a parameter name to represent each element in the collection being iterated over. However, the use of e is not a requirement of the forEach() loop syntax and any valid variable name can be used instead.

The reason e is often used as the parameter name is because it is short for "element" and provides a concise and easily recognizable variable name. Additionally, other programming languages such as JavaScript and C# also commonly use e as a parameter name in similar loop constructs, so using e as a convention can help improve code consistency and readability across languages.

It's worth noting that while e is a common convention, it's important to choose a variable name that is descriptive of the element being iterated over and makes sense within the context of the code being written.